sources/api/bootstrap-containers/src/main.rs

```diff
@@ -0,0 +1,730 @@
+        // Only enable the unit, since it will be started once systemd reaches the `configured`
+        // target
+        command(SYSTEMCTL_BIN, &["enable", &self.unit])
+    }
+
+    fn disable_and_stop(&self) -> Result<()> {
+        command(SYSTEMCTL_BIN, &["disable", &self.unit, "--now"])
+    }
+}
+
+/// Wrapper around process::Command that adds error checking.
+fn command<I, S>(bin_path: &str, args: I) -> Result<()>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let mut command = Command::new(bin_path);
+    command.args(args);
+    let output = command
+        .output()
+        .context(error::ExecutionFailure { command })?;
+
+    trace!("stdout: {}", String::from_utf8_lossy(&output.stdout));
+    trace!("stderr: {}", String::from_utf8_lossy(&output.stderr));
+
+    ensure!(
+        output.status.success(),
+        error::CommandFailure {
+            bin_path,
+            output,
+            code: output.status.code()
+        }
+    );
+    Ok(())
+}
+
+/// Handles the `create-containers` subcommand
+async fn create_containers<P>(socket_path: P) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    let mut failed = 0usize;
+    let bootstrap_containers = get_bootstrap_containers(socket_path).await?;
+    for (name, container_details) in bootstrap_containers.iter() {
+        // Continue to handle other bootstrap containers if we fail one
+        if let Err(e) = handle_bootstrap_container(name, container_details) {
+            failed += 1;
+            error!("Failed to handle bootstrap container '{}': {}", &name, e);
+        }
+    }
+
+    ensure!(
+        failed == 0,
+        error::ManageContainersFailed {
+            failed,
+            tried: bootstrap_containers.len()
+        }
+    );
+
+    Ok(())
+}
+
+/// Handles the `mark-bootstrap` subcommand
+async fn mark_bootstrap<P>(args: MarkBootstrapArgs, socket_path: P) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    let container_id: &str = args.container_id.as_ref();
+    let mode = args.mode.as_ref();
+    info!("Mode for '{}' is '{}'", container_id, mode);
+
+    // When 'mode' is 'once', the containers is marked as 'off' once it
+    // finishes. This guarantees that the the container is only started in
+    // the first boot.
+    if mode != "always" {
+        let formatted = format!("settings.bootstrap-containers.{}.mode", container_id);
+        let key =
+            Key::new(KeyType::Data, &formatted).context(error::KeyFormat { key: formatted })?;
+        let value = serialize_scalar(&"off".to_string()).context(error::Serialize)?;
+
+        let mut map = HashMap::new();
+        map.insert(key, value);
+        let settings: model::Settings = datastore::deserialization::from_map(&map)
+            .context(error::SettingsDeserialize { settings: map })?;
+
+        info!("Turning off container '{}'", container_id);
+        apiclient::set::set(socket_path, &settings)
+            .await
+            .context(error::Set)?;
+    }
```

Comment:
It might be worth adding a comment that this causes another, possibly concurrent invocation of `bootstrap-containers` after the API setting finalizes, which is probably safe but could be fun to debug if not.

Reply:
Done